engine/server/gc.go

```diff
@@ -143,52 +140,65 @@ func getGCPolicy(cfg config.GCConfig, root string) []bkclient.PruneInfo {
-			All:          rule.All,
-			KeepBytes:    rule.KeepBytes.AsBytes(root),
-			KeepDuration: rule.KeepDuration.Duration,
+			Filter:        rule.Filters,
+			All:           rule.All,
+			KeepDuration:  rule.KeepDuration.Duration,
+			ReservedSpace: rule.ReservedSpace.AsBytes(dstat),
+			MaxUsedSpace:  rule.MaxUsedSpace.AsBytes(dstat),
+			MinFreeSpace:  rule.MinFreeSpace.AsBytes(dstat),
 		})
 	}
 	return out
 }
 
-func defaultGCPolicy(keep config.DiskSpace) []config.GCPolicy {
-	if keep == (config.DiskSpace{}) {
-		keep = config.DiskSpace{Percentage: DefaultDiskSpacePercentage}
-	}
-	return []config.GCPolicy{
-		// if build cache uses more than 512MB delete the most easily reproducible data after it has not been used for 2 days
-		{
-			Filters:      []string{"type==source.local,type==exec.cachemount,type==source.git.checkout"},
-			KeepDuration: config.Duration{Duration: time.Duration(48) * time.Hour}, // 48h
-			KeepBytes:    config.DiskSpace{Bytes: 512 * 1e6},                       // 512MB
-		},
-		// remove any data not used for 60 days
-		{
-			KeepDuration: config.Duration{Duration: time.Duration(60) * 24 * time.Hour}, // 60d
-			KeepBytes:    keep,
-		},
-		// keep the unshared build cache under cap
-		{
-			KeepBytes: keep,
-		},
-		// if previous policies were insufficient start deleting internal data to keep build cache under cap
-		{
-			All:       true,
-			KeepBytes: keep,
-		},
+func getDefaultGCPolicy(cfg config.GCConfig, root string) bkclient.PruneInfo {
+	// the last policy is the default one
+	policies := getGCPolicy(cfg, root)
+	return policies[len(policies)-1]
+}
+
+func defaultGCPolicy(cfg config.GCConfig, dstat disk.DiskStat) []config.GCPolicy {
+	if cfg.IsUnset() {
+		cfg.GCReservedSpace = cfg.GCKeepStorage
+	}
+	if cfg.IsUnset() {
+		// use our own default caps, so we can configure the limits ourselves
+		cfg = DetectDefaultGCCap(dstat)
 	}
+	return config.DefaultGCPolicy(cfg, dstat)
 }
 
-func getGCKeepBytesFromConfig(keep config.DiskSpace, root string) int64 {
-	if keep == (config.DiskSpace{}) {
-		keep = config.DiskSpace{Percentage: DefaultDiskSpacePercentage}
+func DetectDefaultGCCap(dstat disk.DiskStat) config.GCConfig {
+	reserve := config.DiskSpace{Percentage: diskSpaceReservePercentage}
+	if reserve.AsBytes(dstat) > diskSpaceReserveBytes {
+		reserve = config.DiskSpace{Bytes: diskSpaceReserveBytes}
+	}
+	max := config.DiskSpace{Percentage: diskSpaceMaxPercentage}
+	if max.AsBytes(dstat) > diskSpaceMaxBytes {
+		max = config.DiskSpace{Bytes: diskSpaceMaxBytes}
+	}
+	return config.GCConfig{
+		GCReservedSpace: reserve,
+		GCMinFreeSpace:  config.DiskSpace{Percentage: diskSpaceFreePercentage},
+		GCMaxUsedSpace:  max,
 	}
-	return keep.AsBytes(root)
 }
+
+const (
+	diskSpaceReservePercentage int64 = 10
+	diskSpaceReserveBytes      int64 = 10 * 1e9 // 10GB
+	diskSpaceFreePercentage    int64 = 20
+	diskSpaceMaxPercentage     int64 = 80
+	diskSpaceMaxBytes          int64 = 200 * 1e9 // 200GB
+)
```

Comment:
The old policy is very simple - use 75% of the disk.

The new policy is more complex:
- Always allow use of 10% or 10GB (whichever is smaller, to handle disks that are very small, like docker-desktop's defaults) - dagger's cache will never be forced to reduce below this amount
- Never consume more than 80% or 200GB (whichever is smaller again) - dagger's cache will also be gc-ed at this amount
- Try and keep 20% of the disk free at any point - but never below the reserved storage of 10%/10GB.

I *think* this makes sense? It does mean than on a huge disk, like a 1TB one, we won't use more than 200GB of storage - I think that's the right call for a local dev experience? For CI, I'm not sure, would appreciate any inputs @marcosnils @sipsma.

Reply:
Yeah, makes sense - I've reverted the default to 75% usage, without the cap - if we do that, we can do it separately.

I'm working on a new config file mechanism as well, which should give us more ownership over our config, and will help us more easily configure the defaults (so we don't have to keep pointing to the buildkit docs).